### PR TITLE
[RLlib] Tests for bandit convergence and solving cov matrix problem

### LINKED ISSUE #29543: [RLlib] MultivariateNormal distributions in bandits do not actually update sample variance
### What happened + What you expected to happen

For the conclusion first, it looks like BanditLinUCB and BanditLinTS are both broken because the variance of the parameter estimates appear to be fixed to 1 at initialization. 

Investigation ahead!

A teammate and I were experimenting with using BanditLinTS against a toy data example we created with arms with fixed reward ratios.

As we ran the experiment, we found that the internal estimates of the arms looked correct. That is to say, as time went on, the covariance matrix (just a single value) shrank, and the value of theta was approaching the correct value.

Still, we were super surprised to find that for some reason, the model wasn't really reliably picking one arm over another.

With that in mind, I tried hitting up the internal distributions of the arms directly to see what the sample function returns. Turns out the means are correct, but empirically the variances for both arms are 1.

Screen Shot 2022-10-21 at 1 27 18 AM

I figured that was really strange, so I messed around with initializing the MultivariateNormal distributions on my own but couldn't reproduce the bug.

I then went and checked the `.__dict__` attribute of the arms returned and noticed something weird. There was a hidden `_unbroadcasted_scale_tril` variable that's 1!

```
{'covariance_matrix': tensor([[0.0002]]),
 'loc': tensor([0.3006]),
 '_batch_shape': torch.Size([]),
 '_event_shape': torch.Size([1]),
 '_unbroadcasted_scale_tril': tensor([[1.]])}
```
That matched my observation.

Looking into torch's source code, it looks like they actually use the `_unbroadcasted_scale` parameter for their sampling: https://github.com/pytorch/pytorch/blob/master/torch/distributions/multivariate_normal.py#L210

So, my guess was that covariance does update, but it doesn't change the internal private variable. And that was correct!

Screen Shot 2022-10-21 at 2 19 50 AM

Lastly, I noticed that the covariance is actually [passed through a cholesky decomp ](https://github.com/pytorch/pytorch/blob/master/torch/distributions/multivariate_normal.py#L155) when setting that internal variable.

When I repeated the experiment by setting the scale_tril variable instead, the update was indeed reflected in the internal state! Notice how the `_unbroadcasted_scale_tril` variable goes from tiny to huge.

Screen Shot 2022-10-21 at 2 27 48 AM

So, it looks like `torch.linalg.cholesky` screws up the whole update propagation thing and you're stuck with the variance you initialized at, which is 1

If you've got two arms whose true values are within 1 of each other (i.e. tons of bernoulli variables) you aren't going to pick a good arm much more than a bad one.

Perhaps even worse, for an experiment where you've got large expected values for the arms that also have variances large relative to 1 (e.g, two arms with true means of 1000, and 1050, true variances of 100 each) You could end up in a situation where you get an slightly unlucky draw on that 1050 arm which plops your mean estimate for it at 950, and then you could just keep drawing from the 1000-valued arm and never get a sample mean near enough to 950 for your fixed 1-variance uncertainties to overlap.

### Versions / Dependencies

Python3.8,
 ray==2.0.0,
torch==1.12.1
Ubuntu 18.04.6 LTS

### Reproduction script
```python
import gym
import numpy as np
from gym.spaces import Discrete, Box
from matplotlib import pyplot as plt
from ray.rllib.algorithms.bandit.bandit import BanditLinTS
from ray.rllib.env import EnvContext
TRAINING_ITERATIONS = 100
ARM_COUNT = 2
BETTER_ARM_PROBABILITY = 0.3
ACTION_TO_REWARD_PROBABILITY = {
    0: 0.1,
    1: BETTER_ARM_PROBABILITY
}
class NonContextualBanditEnv(gym.Env):
    def __init__(self, config: EnvContext):
        self.action_space = Discrete(ARM_COUNT)
        # actually no features - can remove?
        self.observation_space = Box(0.0, 1.0, shape=(1,), dtype=np.float32)
    def reset(self):
        return [1.0]
    def step(self, action):
        reward = np.random.binomial(1.0, ACTION_TO_REWARD_PROBABILITY[action])
        return [1.0], reward, True, {"regret": BETTER_ARM_PROBABILITY - ACTION_TO_REWARD_PROBABILITY[action]}
    
if __name__ == "__main__":
    """
    Test how well the model converges to the better arm that has a higher reward probability.
    """
    algo = BanditLinTS(env=NonContextualBanditEnv, config={'framework': 'torch'})
    policy = algo.get_policy()
    model = policy.model
    arm_means = []
    training_results = []
    reward_per_round = []
    # Log progress for each training iteration
    for i in range(TRAINING_ITERATIONS):
        training_results.append(algo.train())
        arm_means.append([float(model.arms[i].theta) for i in range(ARM_COUNT)])
        reward_per_round.append(sum([algo.compute_single_action(observation=[1.0]) for _ in range(100)]))

print(model.arms[1].dist)
n = 100000
samples = [float(model.arms[1].dist.sample()) for _ in range(n)]
print(np.mean(samples), np.var(samples))
print(model.arms[1].dist.__dict__)
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by avnishn**
The bandit algos had a problem where the cov matrix for the arms does not propagate updates to the multivariate normal distribution, every time the cov matrix is updated. This is because the distribution is meant to be made but not updated over time. The cov matrix used to instantiate the distribution in both torch/tf is used to computed some scale_tril matrix which is used to parameterize the distribution. That scale_tril matrix doesn't get updated when the cov matrix is updated.

The proposed solution is to recreate the distribution every time the cov matrix is updated. 

## Related issue number

closes #29543
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/bandit/tests/test_bandits.py [hunk 1]
```diff
@@ -64,6 +72,85 @@ def test_bandit_lin_ucb_compilation(self):
                 self.assertTrue(results["episode_reward_mean"] == 10.0)
                 trainer.stop()
 
+    def test_bandit_convergence(self):
+        TRAINING_ITERATIONS = 50
+        ARM_COUNT = 2
+        BETTER_ARM_PROBABILITY = 0.3
+        ACTION_TO_REWARD_PROBABILITY = {0: 0.1, 1: BETTER_ARM_PROBABILITY}
+
+        class NonContextualBanditEnv(gym.Env):
```
Comment (jonastim) [inline rllib/algorithms/bandit/tests/test_bandits.py:81]: Cleaned this up a little in the meantime on our end

```
ARM_COUNT = 2
GOOD_ARM_REWARD_PROBABILITY = 0.3
DUMMY_OBSERVATION = 0
REWARD = 1.0

REWARD_PROBABILITY_FOR_ACTION = {
    0: 0.1,                             # bad arm
    1: GOOD_ARM_REWARD_PROBABILITY      # good arm
}


class NonContextualBanditEnv(gym.Env):
    def __init__(self, config: EnvContext):
        self.action_space = Discrete(ARM_COUNT) # bad_arm / good_arm

        # TODO actually no features - can this be removed?
        self.observation_space = Discrete(1)

    def reset(self):
        return DUMMY_OBSERVATION

    def step(self, action):
        reward = np.random.binomial(REWARD, REWARD_PROBABILITY_FOR_ACTION[action])
        return DUMMY_OBSERVATION, reward, True, {"regret": GOOD_ARM_REWARD_PROBABILITY - REWARD_PROBABILITY_FOR_ACTION[action]}
```

and in a different script I use a more flexible implementation for a configurable number of arms
```
DUMMY_OBSERVATION = 0
REWARD = 1.0

class NonContextualBanditEnv(gym.Env):
    def __init__(self, config: EnvContext):
        self.reward_probability_for_action = config['reward_probability_for_action']
        self.highest_reward_probability = max(self.reward_probability_for_action.values())

        self.action_space = Discrete(len(self.reward_probability_for_action)) 

        # TODO actually no features - can this be removed?
        self.observation_space = Discrete(1)

    def reset(self):
        return DUMMY_OBSERVATION

    def step(self, action):
        reward = np.random.binomial(REWARD, self.reward_probability_for_action[action])
        return DUMMY_OBSERVATION, reward, True, {"regret": self.highest_reward_probability - self.reward_probability_for_action[action]}
```

File: rllib/algorithms/bandit/tests/test_bandits.py [hunk 2]
```diff
@@ -64,6 +69,69 @@ def test_bandit_lin_ucb_compilation(self):
...
                 trainer.stop()
 
+    def test_bandit_convergence(self):
+        # test whether in a simple bandit environment, the bandit algorithm
+        # distribution converge to the optimal distribution empirically
+
+        std_threshold = 0.1
+        best_arm_prob = 0.5
+
+        class NonContextualBanditEnv(gym.Env):
```
Comment (gjoliver) [inline rllib/algorithms/bandit/tests/test_bandits.py:79]: can you please move this outside of the test case to the top level of the file?
there doesn't seem to be a good reason for this to be an inner class. just let the __init__ function take best_arm_prob as a parameter.

**Review by gjoliver — COMMENTED** [review]
nice fix!! 😓

**Comment by avnishn** [comment]
Ah so that's how it's supposed to work. The more exploited arm should have a lower variance. 

Thanks for pulling this through man 🥳😘

File: rllib/algorithms/bandit/tests/test_bandits.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/algorithms/bandit/tests/test_bandits.py:79]: these are not shared with any other tests tho. My convention is usually to add stuff to the top if they are shared by at least more than 1 unittest.

Comment (avnishn) [reply rllib/algorithms/bandit/tests/test_bandits.py:79]: Thinking back on it, inner classes are probably only appropriate if we are auto generating a class via a parameterized function.

Comment (kouroshHakha) [reply rllib/algorithms/bandit/tests/test_bandits.py:79]: Oh i thought @gjoliver was talking about the constant numbers. I totally agree that this class should go outside of the unittest for better readability.

Comment (avnishn) [reply rllib/algorithms/bandit/tests/test_bandits.py]: Oh lmao guess we're all on the same page then.

**Review by gjoliver — APPROVED** [review]

**Comment by gjoliver** [comment]
lint is failing ... :(
can you update black version and try again?

**Comment by kouroshHakha** [comment]
@gjoliver fixed.
